chia/server/upnp.py

```diff
@@ -7,16 +7,33 @@
 
 
 log = logging.getLogger(__name__)
+try:
+    upnp = miniupnpc.UPnP()
+    upnp.discoverdelay = 30
+    upnp.discover()
+    upnp.selectigd()
+except Exception as e:
+    log.info("UPnP failed. This is not required to run chia, but it allows incoming connections from other peers.")
+    log.info(e)
 
 
 def upnp_remap_port(port) -> None:
     log.info(f"Attempting to enable UPnP (open up port {port})")
     try:
-        upnp = miniupnpc.UPnP()
-        upnp.discoverdelay = 30
-        upnp.discover()
-        upnp.selectigd()
+        global upnp
+        upnp.deleteportmapping(port, "TCP")
```

Comment:
these calls also have to release the coroutine scheduler, but it looks like they're blocking now, which would stall the main thread and all handling of peer traffic.

Reply:
How is this blocking now, but not before?

Reply:
Please find out whether they are blocking or not. Also, if they are blocking, please find out whether they correctly release the GIL. If not, we may have to do something more complicated, perhaps patch the miniupnp module.

I suspect they are blocking because they depend on each other. You can't map the port until you have discovered SSDP devices, and you then have to select a gateway device before you can map, and so on.

I would expect these calls to have to be posted to the thread as well, for two reasons.

1. To ensure they always happen *after* we're done with `upnp.discover()` and `upnp.selectgd()`
2. To not block the main thread

Reply:
The calls are blocking (it's noticeable just running the methods), but they do all appear to release the GIL, thankfully. See https://github.com/miniupnp/miniupnp/blob/master/miniupnpc/miniupnpcmodule.c#L399